content/blog/2020-12-20-loop-flatten/index.md

```diff
@@ -0,0 +1,199 @@
+
+As we know, loops are usually the heavy part during execution, which consumes most of the runtime. Thus, many strategies have been applied to speed up the execution time of the loops. For example, Loop Invariant Code Motion (LICM) removes the loop invariant instructions out of the loop to avoid perform useless instructions. For more details about all available loop optimizations, please check out [this link](https://llvm.org/docs/Passes.html). And for the explanation of loop optimizing terminologies, please check out [this link](https://llvm.org/docs/TransformMetadata.html).
+
+## Motivation
+As we mentioned before, with the multi-core compute architecture, we would like to leverage the parallelism of the code to get better performance. As for some loops, there is no data dependency around multiple iterations, which is a good fit for loop unrolling optimization. Specifically, multiple loop iterations can be executed independently at the same time. Therefore, in this section, we will introduce the loop unrolling optimizations and introduce the motivation of adding the loop flattening optimization.
+
+1. Loop unrolling<br/>
+In LLVM, this optimization unrolls the loop with the given unroll parameters. To be more specific, it will transform the following piece of code:
+
+```javascript
+for (int i = 0; i < n; i++) { // original loop
+  Stmt(i);
+}
+```
+
+Into this piece of code:
+
+```javascript
+for (int i = 0; i < n; i+=4) { // unrolledloop
+  Stmt(i);
+  Stmt(i+1);
+  Stmt(i+2);
+  Stmt(i+3);
+}
+
+for (int i = 0; i < m; i++) { // remainder loop
+  Stmt(i);                            // m = n % 4 
+}
+```
+
+2. Loop unroll-and-jam<br/>
+In LLVM, this optimization unrolls the nested loop with the given unroll parameters. To be more specific, it will transform the following piece of code:
+
+```javascript
+for (int i = 0; i < n; i++) { // original outer loop
+  Outer_Stmt(i);
+  for (int j = 0; j < m; j++) { // original inner loop
+    Inner_Stmt(i);
+  }
+}
+```
+Into this piece of code:
+
+```javascript
+for (int i = 0; i < n; i+= 2) { // unrolled outer loop
+  Outer_Stmt(i);
+  Outer_Stmt(i+1);
+  for (int j = 0; j < m; j+= 2) { // unrolled inner loop
+    Inner_Stmt(i);
+    Inner_Stmt(i+1);
+  }
+}
+```
+
+3. Loop flattening<br/>
+As illustrated by its name, this optimization will flatten the nested loop into a single-level loop. 
+To be more specific, it will transform the following piece of code:
+```javascript
+  for (int i = 0; i < n; i++) { // unflattened nested loop
+    for (int j = 0; j < m; j++) {
+      Stmt(f(i, j))
+  }
+}
+```
+Into this piece of code:
+
+```javascript
+for (int i = 0; i < n*m; i+= 2) { // flattened loop
+  Stmt(f(i))

```

Comment:
It's not entirely clear what's going on here—`f` used to be a two-argument function, and now it's a one-argument function. Did something have to change inside `f`?